src/block-data-providers/bitcoin-core/provider.ts

```diff
@@ -0,0 +1,278 @@
+import axios from 'axios';
+
+@Injectable()
+export class BitcoinCoreProvider
+    extends BaseBlockDataProvider
+    implements OnApplicationBootstrap
+{
+    protected readonly logger = new Logger(BitcoinCoreProvider.name);
+    protected readonly operationStateKey = 'bitcoincore-operation-state';
+    private isSyncing = false;
+    private config: BitcoinCoreConfig;
+
+    public client: Client;
+
+    public constructor(
+        private configService: ConfigService,
+        indexerService: IndexerService,
+        operationStateService: OperationStateService,
+    ) {
+        super(indexerService, operationStateService);
+        this.config = this.configService.get<BitcoinCoreConfig>('bitcoinCore');
+    }
+
+    async onApplicationBootstrap() {
+        const getState = await this.getState();
+        if (getState) {
+            this.logger.log(
+                `Restoring state from previous run: ${JSON.stringify(
+                    getState,
+                )}`,
+            );
+        } else {
+            this.logger.log('No previous state found. Starting from scratch.');
+            const state: BitcoinCoreOperationState = {
+                currentBlockHeight: 0,
+                indexedBlockHeight:
+                    this.configService.get<BitcoinNetwork>('app.network') ===
+                    BitcoinNetwork.MAINNET
+                        ? TAPROOT_ACTIVATION_HEIGHT - 1
+                        : 0,
+            };
+            await this.setState(state);
+        }
+    }
+
+    @Cron(CronExpression.EVERY_10_SECONDS)
+    async sync() {
+        if (this.isSyncing) return;
+        this.isSyncing = true;
+
+        const state = await this.getState();
+        if (!state) {
+            throw new Error('State not found');
+        }
+        const tipHeight = await this.getTipHeight();
+        if (tipHeight <= state.indexedBlockHeight) {
+            this.logger.debug(
+                `No new blocks found. Current tip height: ${tipHeight}`,
+            );
+            this.isSyncing = false;
+            return;
+        }
+
+        let height = state.indexedBlockHeight + 1;
+
+        for (height; height <= tipHeight; height++) {
+            const transactions = await this.processBlock(height);
+            for (const transaction of transactions) {
+                const { txid, vin, vout, blockHeight, blockHash } = transaction;
+                await this.indexTransaction(
+                    txid,
+                    vin,
+                    vout,
+                    blockHeight,
+                    blockHash,
+                );
+            }
+            state.indexedBlockHeight = height;
+            await this.setState(state);
+        }
+        this.isSyncing = false;
+    }
+
+    async getTipHeight(): Promise<number> {
+        try {
+            const body = {
+                method: 'getblockcount',
+                params: [],
+                ...this.createPartialRPCBody(),
```

Comment:
If we have to call `this.createPartialRPCBody()` everytime before calling `this.request()` then it's better to get rid of this function and move it inside `request()`